In [56]:
import re

import numpy as np
import pandas as pd

In [57]:
hh_set_ini = pd.read_csv('data/dst-3.0_16_1_hh_database.csv', sep=';')

----

### Задание 2.1 ###

Прочитайте данные с помощью библиотеки Pandas. Перед чтением обратите внимание на разделитель внутри файла.
Чему равна размерность таблицы? Введите её в виде кортежа (число строк, число столбцов).

In [58]:
hh_set_ini.shape

(44744, 12)

----

### Задание 2.2 ###

Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.
В каком формате представлен порядок слов в признаке «Пол, возраст»?

***Ответ - <Пол , возраст , дата рождения>***

In [59]:
hh_set_ini.head(3)

Пол, возраст          ЗП  \
0    Мужчина ,  39 лет , родился 27 ноября 1979  29000 руб.   
1     Мужчина ,  60 лет , родился 20 марта 1959  40000 руб.   
2  Женщина ,  36 лет , родилась 12 августа 1982  20000 руб.   

  Ищет работу на должность:  \
0   Системный администратор   
1      Технический писатель   
2                  Оператор   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   
2  Тверь , не готова к переезду , не готова к ком...   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   
2                                   полная занятость   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   
2                                        полный день   

                                         Опыт работы  \
0  Опыт работы 16 лет 10 месяцев  Август 2010 — п...   
1  Опыт работы 19 лет 5 месяцев  Январь 2000 — по...   
2  Опыт работы 10 лет 3 месяца  Октябрь 2004 — Де...   

  Последнее/нынешнее место работы  \
0         МАОУ "СОШ № 1 г.Немана"   
1    Временный трудовой коллектив   
2                    ПАО Сбербанк   

                       Последняя/нынешняя должность  \
0                           Системный администратор   
1  Менеджер проекта, Аналитик, Технический писатель   
2                               Кассир-операционист   

                                   Образование и ВУЗ Обновление резюме  \
0  Неоконченное высшее образование 2000  Балтийск...  16.04.2019 15:59   
1  Высшее образование 1981  Военно-космическая ак...  12.04.2019 08:42   
2  Среднее специальное образование 2002 Профессио...  16.04.2019 08:35   

                             Авто  
0  Имеется собственный автомобиль  
1                      Не указано  
2                      Не указано

----

### Задание 2.3 ###

Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.
Какой тип данных имеют все столбцы? Введите его в поле для ответа.

***Ответ - object***

In [60]:
hh_set_ini.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

----

### Задание 2.4 ###

Обратите внимание на информацию о числе непустых значений.
Выберите все признаки, в которых есть пропуски в данных:

***Ответ***
- Последняя/нынешняя должность
- Опыт работы
- Последнее/нынешнее место работы

----

### Задание 2.5 ###

Выведите основную статистическую информацию о столбцах.
Сколько уникальных значений содержится в столбце «Опыт работы»?

***Ответ - 44413***

Какая из искомых должностей самая распространённая?

***Ответ - Системный администратор***

In [61]:
hh_set_ini.describe(include=['object'])

Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График                                        Опыт работы  \
count         44744                                              44576   
unique           47                                              44413   
top     полный день  Опыт работы 10 лет 8 месяцев  Апрель 2018 — по...   
freq          22727                                                  3   

                          Последнее/нынешнее место работы  \
count                                               44743   
unique                                              30214   
top     Индивидуальное предпринимательство / частная п...   
freq                                                  935   

       Последняя/нынешняя должность  \
count                         44742   
unique                        16927   
top         Системный администратор   
freq                           2062   

                                        Образование и ВУЗ Обновление резюме  \
count                                               44744             44744   
unique                                              40148             18838   
top     Высшее образование 1987 Военный инженерный Кра...  07.05.2019 09:50   
freq                                                    4                25   

              Авто  
count        44744  
unique           2  
top     Не указано  
freq         32268

----

###  Задание 3.1 ###

Сколько соискателей имеет средний уровень образования (школьное образование)?

***Ответ - 559***

In [69]:
display(hh_set_ini['Образование и ВУЗ'])

def get_graduation(string_):
    list_ = string_.split(' ')
    if list_[1] == 'образование':
        return list_[0]
    else:
        return list_[0] + ' ' + list_[1]
    return None


hh_set_ini['Образование'] = hh_set_ini['Образование и ВУЗ'].apply(get_graduation)
print('Уникальных значений образования -', hh_set_ini['Образование'].nunique())
display(hh_set_ini['Образование'].value_counts())


0        Неоконченное высшее образование 2000  Балтийск...
1        Высшее образование 1981  Военно-космическая ак...
2        Среднее специальное образование 2002 Профессио...
3        Высшее образование 2002  Саратовский государст...
4        Высшее образование 2015 Кгу Психологии и педаг...
                               ...                        
44739    Высшее образование 2015  Московский гуманитарн...
44740    Высшее образование (Бакалавр) 2016  Воронежски...
44741    Высшее образование 2000  Южно-Уральский госуда...
44742    Высшее образование 2015  Тамбовский государств...
44743    Высшее образование 1997  Южно-Российский госуд...
Name: Образование и ВУЗ, Length: 44744, dtype: object

Уникальных значений образования - 4


Высшее                 33863
Среднее специальное     5765
Неоконченное высшее     4557
Среднее                  559
Name: Образование, dtype: int64

In [70]:
def get_graduation_year(string_):
    return int(re.search(r'\d{4}', string_).group(0))

hh_set_ini['Год выпуска'] = hh_set_ini['Образование и ВУЗ'].apply(get_graduation_year)
print('Уникальных значений года выпуска -', hh_set_ini['Год выпуска'].nunique())
display(hh_set_ini['Год выпуска'].value_counts())
display(hh_set_ini.info())

Уникальных значений года выпуска - 62


2015    3004
2013    2756
2014    2705
2012    2674
2011    2654
        ... 
2028       3
2029       2
1954       1
1969       1
1972       1
Name: Год выпуска, Length: 62, dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
 12  Образование                     

None

----

### Задание 3.2 ###

Сколько процентов женских резюме представлено в наших данных?

(Ответ округлить до сотых)

***Ответ - 19.07***

Чему равен средний возраст соискателей? Ответ округлите до десятых.

***Ответ - 32.2***

In [78]:
display(hh_set_ini['Пол, возраст'])

hh_set_ini['Пол'] = hh_set_ini['Пол, возраст'].apply(lambda x: x[0])
display(hh_set_ini['Пол'].value_counts())
print(round(hh_set_ini[hh_set_ini['Пол'] == 'Ж'].shape[0] / hh_set_ini['Пол'].shape[0] * 100, 2))

0          Мужчина ,  39 лет , родился 27 ноября 1979
1           Мужчина ,  60 лет , родился 20 марта 1959
2        Женщина ,  36 лет , родилась 12 августа 1982
3            Мужчина ,  38 лет , родился 25 июня 1980
4           Женщина ,  26 лет , родилась 3 марта 1993
                             ...                     
44739      Мужчина ,  30 лет , родился 17 января 1989
44740        Мужчина ,  27 лет , родился 5 марта 1992
44741    Женщина ,  48 лет , родилась 26 декабря 1970
44742     Мужчина ,  24 года , родился 6 октября 1994
44743      Мужчина ,  38 лет , родился 25 апреля 1980
Name: Пол, возраст, Length: 44744, dtype: object

М    36211
Ж     8533
Name: Пол, dtype: int64

19.07


In [82]:
def get_age(string_):
    age_str = string_.split(',')[1].strip()
    return int(re.search(r'\d+', age_str).group(0))

##### hh_set_ini['Возраст'] = hh_set_ini['Пол, возраст'].apply(lambda x: x.split(',')[1].strip())
hh_set_ini['Возраст'] = hh_set_ini['Пол, возраст'].apply(get_age)
print(round(hh_set_ini['Возраст'].mean(), 2))

32.2


----

### Задание 3.3 ###

Чему равен медианный опыт работы (в месяцах) в таблице?

----

### Задание 3.4 ###

Сколько процентов соискателей живут в Санкт-Петербурге?

(Ответ округлитm до целого)

***Ответ -***

Сколько процентов соискателей готовы одновременно и к переездам, и к
командировкам?

(Ответ округлить до целого)

***Ответ -***

----
### Задание 3.5 ###

Сколько людей ищут проектную работу и волонтёрство (в обоих столбцах стоит
`True`)?

***Ответ -***

Сколько людей хотят работать вахтовым методом и с гибким графиком (в обоих
столбцах стоит `True`)?

***Ответ -***

----
### Задание 3.6 ###

Для выполнения этого задания у вас будет 5 попыток. После того как вы дадите
правильный ответ ИЛИ истратите все попытки, вы можете нажать на кнопку «Показать
ответ»: в конце решения вы увидите ссылку на эталонный ответ.

Чему равна желаемая медианная заработная плата соискателей в нашей таблице (в
рублях)?

Ответ приведите в тысячах рублей и округлите до целого. Например, если у вас
получилась медианная з/п 42 127, в качестве ответа укажите 42.

----

### Задание 4.1 ###

Чему равно модальное значение возраста соискателей?

----

### Задание 4.2 ###

Чему равен максимальный опыт работы (в месяцах)?

----

### Задание 4.3 ###

Определите по графику, сколько соискателей требуют заработную плату выше 1 миллиона рублей.

----

###  Задание 4.4 ###

Какая категория образования оплачивается выше всех по медианному показателю?

----

### Задание 4.5 ###

В каком городе (категории городов) зафиксирован наибольший показатель желаемой заработной платы (~924 тысячи рублей)?

----

### Задание 4.6 ###

Чему равна желаемая медианная заработная плата соискателей, готовых и к переезду, и к командировкам? Ответ приведите в тысячах, округлив до целого (например, 45).

----

### Задание 4.7 ###

Для какой категории образования наблюдается самый быстрый карьерный рост (то есть интенсивность роста заработной платы наибольшая)?

----

### Задание 4.8 ###

Сколько точек лежат строго выше построенной прямой? Так как точки могут сливаться с прямой, то для ответа на этот вопрос проще всего воспользоваться интерактивной визуализацией либо получить таблицу с наблюдениями, лежащими выше прямой с помощью фильтрации.